presto-main/src/main/java/io/prestosql/sql/planner/LogicalPlanner.java

```diff
@@ -340,29 +340,28 @@ private RelationPlan createInsertPlan(Analysis analysis, Insert insertStatement)
 
         TableMetadata tableMetadata = metadata.getTableMetadata(session, insert.getTarget());
 
-        List<ColumnMetadata> visibleTableColumns = tableMetadata.getColumns().stream()
-                .filter(column -> !column.isHidden())
-                .collect(toImmutableList());
-        List<String> visibleTableColumnNames = visibleTableColumns.stream()
-                .map(ColumnMetadata::getName)
-                .collect(toImmutableList());
-
         RelationPlan plan = createRelationPlan(analysis, insertStatement.getQuery());
 
         Map<String, ColumnHandle> columns = metadata.getColumnHandles(session, insert.getTarget());
         Assignments.Builder assignments = Assignments.builder();
+        boolean supportsMissingColumnsOnInsert = metadata.supportsMissingColumnsOnInsert(session, insert.getTarget());
+        ImmutableList.Builder<ColumnHandle> columnsToBeInserted = ImmutableList.builder();
+        ImmutableList.Builder<ColumnMetadata> visibleColumnsBuilder = ImmutableList.builder();
+
         for (ColumnMetadata column : tableMetadata.getColumns()) {
             if (column.isHidden()) {
                 continue;
             }
             Symbol output = symbolAllocator.newSymbol(column.getName(), column.getType());
             int index = insert.getColumns().indexOf(columns.get(column.getName()));
             boolean missingColumn = index < 0;
-            if (missingColumn) {
+            if (missingColumn && !supportsMissingColumnsOnInsert) {
                 Expression cast = new Cast(new NullLiteral(), toSqlType(column.getType()));
                 assignments.put(output, cast);
+                columnsToBeInserted.add(columns.get(column.getName()));
+                visibleColumnsBuilder.add(column);
             }
-            else {
+            else if (!missingColumn) {
```

Comment:
I don't like this `if` here. It suggests that there is a missing `else` that it is not handled. As I understand this code, here it should be plain `else`.

Reply:
then we might have a inner if condition which would skip casting null for missing columns